cmake/grpc.cmake

```diff
@@ -11,39 +11,115 @@ unset(_grpc_cxx_standard)
+#
+# Variables:
+#   GRPC_PATCH - name of gRPC patch file.
+#   GRPC_VERSION - version of gRPC being patched.
+#
+# Choices are, in descending order of precedence:
+#  1. ${GRPC_PATCHFILE}
+#  2. grpc-v$(GRPC_VERSION}.patch.in
+#  3. grpc-current.patch.in
+#
+# Returns an empty string if we cannot determine a valid patch file name.
+#-----------------------------------------------------------------------
+function(define_grpc_patch_file PATCHDIR OUTVAR)
+  # Specified patch file
+  if(DEFINED GRPC_PATCHFILE AND NOT GRPC_PATCHFILE STREQUAL "")
+    set(patchfile "${GRPC_PATCHFILE}")
+    if(EXISTS ${PATCHDIR}/${patchfile})
+      set(${OUTVAR} "${patchfile}" PARENT_SCOPE)
+    else()
+      message("----------------------------------------")
+      message("Patch file '${patchfile}' does not exist.")
+      message("Check definition of GRPC_PATCHFILE.")
+      message("Its current value is '${GRPC_PATCHFILE}.'")
+      message("----------------------------------------")
+      message(WARNING "Skipping patch stage")
+      set(${OUTVAR} "" PARENT_SCOPE)
+    endif()
     return()
   endif()
 
-  set(patchfile grpc-v${GRPC_VERSION}.patch.in)
-  set(patchdir ${CMAKE_SOURCE_DIR}/cmake/patches)
+  # Patch file for gRPC version
+  if(DEFINED GRPC_VERSION)
+    set(version_file "grpc-v${GRPC_VERSION}.patch.in")
+    if(EXISTS ${PATCHDIR}/${version_file})
+      set(${OUTVAR} "${version_file}" PARENT_SCOPE)
+      return()
+    endif()
+  endif()
 
-  if(NOT EXISTS ${patchdir}/${patchfile})
-    message(WARNING "Patch file for gRPC version ${GRPC_VERSION} not found")
+  # Current patch file
+  set(patchfile "grpc-current.patch.in")
+  if(EXISTS ${PATCHDIR}/${patchfile})
+    set(${OUTVAR} "${patchfile}" PARENT_SCOPE)
     return()
   endif()
 
-  # Patch the gRPC build script to set the RUNPATH of the installed
-  # Protobuf compiler plugins to the relative paths of the library
-  # directories.
+  # Patch file not found
+  if(DEFINED version_file)
+    message("----------------------------------------")
+    message("Patch file '${version_file}' does not exist.")
+    message("Check definition of GRPC_VERSION.")
+    message("Its current value is '${GRPC_VERSION}'.")
+    message("----------------------------------------")
+    message(WARNING "Skipping patch stage")
+  else()
+    message(STATUS "Skipping patch stage")
+  endif()
+  set(${OUTVAR} "" PARENT_SCOPE)
+endfunction(define_grpc_patch_file)
+
+#-----------------------------------------------------------------------
+# define_grpc_patch_clause()
+# Generates the patch file from the template and returns a PATCH
+# clause for the ExternalProject_Add() command. Returns an empty
+# string if patching is suppressed.
+#-----------------------------------------------------------------------
+function(define_grpc_patch_clause OUTVAR)
+  set(patch_dir ${CMAKE_SOURCE_DIR}/cmake/patches)
+
+  define_grpc_patch_file("${patch_dir}" patch_file)
+  if(patch_file STREQUAL "")
+    return()
```

Comment:
Should we set OUTVAR to empty string before returning here - set(${OUTVAR} "" PARENT_SCOPE)?

Reply:
Fair question. I can't remember if I had specific reason for wanting the value to remain unset. It has the same effect when the ${_patch_clause} substitution is performed.

Reply:
Fixed.